activation/wire/malfeasance_invalid_prev_atx.go

```diff
@@ -0,0 +1,231 @@
+// 3. Both ATXs reference the same previous ATX for the same identity.
+// 4. Both marriage certificates have valid signatures.
+//
+// HINT: this works if the identity that publishes the marriage ATX marries themselves.
+type ProofInvalidPrevAtxV2 struct {
+	// NodeID is the node ID that referenced the same previous ATX twice.
+	NodeID types.NodeID
+
+	// PrevATX is the ATX that was referenced twice.
+	PrevATX types.ATXID
+
+	Proofs [2]InvalidPrevAtxProof
+}
+
+var _ Proof = &ProofInvalidPrevAtxV2{}
+
+func NewInvalidPrevAtxProofV2(
+	db sql.Executor,
+	atx1, atx2 *ActivationTxV2,
+	nodeID types.NodeID,
+) (*ProofInvalidPrevAtxV2, error) {
+	if atx1.ID() == atx2.ID() {
+		return nil, errors.New("ATXs have the same ID")
+	}
+
+	if atx1.MarriageATX == nil && atx1.SmesherID != nodeID {
+		return nil, fmt.Errorf("ATX1 was not created by %s", nodeID)
+	}
+	equivocationSet1 := []types.NodeID{atx1.SmesherID}
+	if atx1.MarriageATX != nil {
+		var err error
+		equivocationSet1, err = identities.EquivocationSetByMarriageATX(db, *atx1.MarriageATX)
+		if err != nil {
+			return nil, fmt.Errorf("get equivocation set for ATX1: %w", err)
+		}
+	}
+	marriageIndex1 := slices.Index(equivocationSet1, nodeID)
+	if marriageIndex1 == -1 {
+		return nil, fmt.Errorf("%s did not contribute to ATX1", nodeID)
+	}
+
+	proof1, err := createInvalidPrevAtxProof(atx1, uint32(marriageIndex1))
+	if err != nil {
+		return nil, fmt.Errorf("proof for atx1: %w", err)
+	}
+
+	if atx2.MarriageATX == nil && atx2.SmesherID != nodeID {
+		return nil, fmt.Errorf("ATX2 was not created by %s", nodeID)
+	}
+	equivocationSet2 := []types.NodeID{atx2.SmesherID}
+	if atx2.MarriageATX != nil {
+		var err error
+		equivocationSet2, err = identities.EquivocationSetByMarriageATX(db, *atx2.MarriageATX)
+		if err != nil {
+			return nil, fmt.Errorf("get equivocation set for ATX2: %w", err)
+		}
+	}
+	marriageIndex2 := slices.Index(equivocationSet2, nodeID)
+	if marriageIndex2 == -1 {
+		return nil, fmt.Errorf("%s did not contribute to ATX2", nodeID)
+	}
+
+	proof2, err := createInvalidPrevAtxProof(atx2, uint32(marriageIndex2))
+	if err != nil {
+		return nil, fmt.Errorf("proof for atx2: %w", err)
+	}
+
+	proof := &ProofInvalidPrevAtxV2{
+		NodeID:  nodeID,
+		PrevATX: types.EmptyATXID,
+		Proofs:  [2]InvalidPrevAtxProof{proof1, proof2},
+	}
+	return proof, nil
+}
+
+func createInvalidPrevAtxProof(atx *ActivationTxV2, marriageIndex uint32) (InvalidPrevAtxProof, error) {
+	prevATXsRootProof, err := prevRootProof(atx)
+	if err != nil {
+		return InvalidPrevAtxProof{}, fmt.Errorf("create proof for previous atx root: %w", err)
+	}
+
+	prevATXProof, err := prevProof(atx.PreviousATXs, marriageIndex)
+	if err != nil {
+		return InvalidPrevAtxProof{}, fmt.Errorf("create proof for previous atx: %w", err)
+	}
+
+	proof := InvalidPrevAtxProof{
+		ATXID: atx.ID(),
+
+		PreviousATXsRoot:      types.Hash32(atx.PreviousATXs.Root()),
+		PreviousATXsRootProof: prevATXsRootProof,
+
+		PrevATXIndex: marriageIndex,
```

Comment:
The previous ATX and marriage indexes mean different things and they are not usually equal.

Reply:
Updated to proof the used prevATX correctly